File: src/com/lonelybytes/swiftlint/SwiftLint.java
package com.lonelybytes.swiftlint;

import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SwiftLint {
    public List<String> executeSwiftLint(@NotNull final String toolPath, @NotNull final String aAction, @NotNull SwiftLintConfig aConfig, @NotNull final String aFilePath) throws IOException, InterruptedException {
        if (aAction.equals("autocorrect")) {
            processAutocorrect(toolPath, aConfig, aFilePath);
        } else {
            if (aConfig.shouldBeLinted(aFilePath, true)) {
                return processAsApp(toolPath, aAction, aConfig, aFilePath);
            }
        }

        return Collections.emptyList();
    }

    private void processAutocorrect(String aToolPath, SwiftLintConfig config, String aFilePath) throws IOException, InterruptedException {
        List<String> params = new ArrayList<>();
        params.add(aToolPath);
        params.add("autocorrect");
        params.add("--no-cache");
        SwiftLintConfig.Config configForFile = config.getConfig(aFilePath);
        if (configForFile != null) {
            params.add("--config");
            params.add(configForFile._file.getAbsolutePath());
        }
        params.add("--path");
        params.add(aFilePath);

        Process process = Runtime.getRuntime().exec(params.toArray(new String[0]));
        process.waitFor();
    }

    @NotNull
    private List<String> processAsApp(@NotNull String toolPath, @NotNull final String aAction, SwiftLintConfig config, @NotNull String aFilePath) throws IOException, InterruptedException {
        List<String> params = new ArrayList<>();
        params.add(toolPath);
        params.add(aAction);
        params.add("--no-cache");
        params.add("--reporter");
        params.add("csv");
        params.add("--path");
        params.add(aFilePath);
        SwiftLintConfig.Config configForFile = config.getConfig(aFilePath);
        if (configForFile != null) {
            params.add("--config");
            params.add(configForFile._file.getAbsolutePath());
        }
        Process process = Runtime.getRuntime().exec(params.toArray(new String[0]));
        process.waitFor();
        return processSwiftLintOutput(process);
    }

    @NotNull
    private List<String> processSwiftLintOutput(Process aProcess) {
        final List<String> outputLines = new ArrayList<>();
        final List<String> errorLines = new ArrayList<>();

        BufferedReader inputStream = new BufferedReader(new InputStreamReader(aProcess.getInputStream()));
        BufferedReader errorStream = new BufferedReader(new InputStreamReader(aProcess.getErrorStream()));

        try {
            String line;
            while ((line = inputStream.readLine()) != null) {
                outputLines.add(line);
            }

            while ((line = errorStream.readLine()) != null) {
                String testLine = line.toLowerCase();
                if (testLine.contains("error:") || testLine.contains("warning:") || testLine.contains("invalid:") || testLine.contains("unrecognized arguments:")) {
                    errorLines.add(line);
                }
            }
        } catch (IOException e) {
            if (!e.getMessage().contains("closed")) {
                Notifications.Bus.notify(new Notification(Configuration.KEY_SWIFTLINT, "Error", "IOException: " + e.getMessage(), NotificationType.INFORMATION));
            }

            e.printStackTrace();
        }

        for (String errorLine : errorLines) {
            if (!errorLine.trim().isEmpty()) {
                Notifications.Bus.notify(new Notification(Configuration.KEY_SWIFTLINT, "Error", "SwiftLint error: " + errorLine, NotificationType.INFORMATION));
            }
        }

        return outputLines;
    }
}

File: src/com/lonelybytes/swiftlint/SwiftLint.kt
package com.lonelybytes.swiftlint

import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.notification.Notifications
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread


class SwiftLint {
    companion object {
        private const val DEBUG_ON = false
        private const val THREAD_WAIT_TIMEOUT:Long = 10000

        fun log(message: String) {
            if (DEBUG_ON) println(message)
        }
    }

    @Throws(IOException::class, InterruptedException::class)
    fun executeSwiftLint(toolPath: String, aAction: String, aFilePath: String, aRunDirectory: File): List<String> {
        return if (aAction == "autocorrect") {
            processAutocorrect(toolPath, aFilePath, aRunDirectory)
            emptyList()
        } else {
            processAsApp(toolPath, aAction, aFilePath, aRunDirectory)
        }
    }

    @Throws(IOException::class, InterruptedException::class)
    fun getSwiftLintRulesList(aToolPath: String): List<String> {
        val params = mutableListOf(aToolPath, "rules")
        val process = Runtime.getRuntime().exec(params.toTypedArray())
        return processSwiftLintOutput(process)
    }

    @Throws(IOException::class, InterruptedException::class)
    private fun processAutocorrect(aToolPath: String, aFilePath: String, aRunDirectory: File) {
        val params = mutableListOf(aToolPath, "autocorrect", "--no-cache", "--path", aFilePath)

        log(" --> # Run: '" + params.joinToString(separator = " ") + "' in '" + aRunDirectory.path + "'")

        val process = Runtime.getRuntime().exec(params.toTypedArray(), emptyArray(), aRunDirectory)
        processSwiftLintOutput(process) // need this, otherwise swiftlint can wait indefinitely in case of lots of output
    }

    @Throws(IOException::class, InterruptedException::class)
    private fun processAsApp(toolPath: String, aAction: String, aFilePath: String, aRunDirectory: File): List<String> {
        val params: MutableList<String> = mutableListOf(toolPath, aAction, "--no-cache", "--reporter", "csv", "--path", aFilePath)

        log(" --> # Run: '" + params.joinToString(separator = " ") + "' in '" + aRunDirectory.path + "'")

        val process = Runtime.getRuntime().exec(params.toTypedArray(), emptyArray(), aRunDirectory)
        return processSwiftLintOutput(process)
    }

    private fun processSwiftLintOutput(aProcess: Process): List<String> {
        log("Started to process swiftlint output...")

        val outputBufferedReader = aProcess.inputStream.bufferedReader(Charset.forName("UTF8"))
        var outputLines: List<String> = arrayListOf()
        val outputThread = thread(true) {
            try {
                outputLines = outputBufferedReader.readLines()
            } catch (e: IOException) {
                if (!e.message!!.contains("closed")) {
                    Notifications.Bus.notify(
                        Notification(Configuration.KEY_SWIFTLINT, "Error", "IOException: " + e.message, NotificationType.INFORMATION)
                    )
                }
                e.printStackTrace()
            }
        }

        val errorBufferedReader = aProcess.errorStream.bufferedReader(Charset.forName("UTF8"))
        var errorLines: List<String> = arrayListOf()
        val errorThread = thread(true) {
            try {
                errorLines = errorBufferedReader.readLines()
                    .filter {
                        val line = it.lowercase()
                        line.contains("error:") || line.contains("warning:") || line.contains("invalid:") || line.contains("unrecognized arguments:")
                    }
            } catch (e: IOException) {
                if (!e.message!!.contains("closed")) {
                    Notifications.Bus.notify(
                        Notification(Configuration.KEY_SWIFTLINT, "Error", "IOException: " + e.message, NotificationType.INFORMATION)
                    )
                }
                e.printStackTrace()
            }
        }

        outputThread.join(THREAD_WAIT_TIMEOUT)
        errorThread.join(THREAD_WAIT_TIMEOUT)

        for (errorLine in errorLines) {
            if (errorLine.trim { it <= ' ' }.isNotEmpty()) {
                Notifications.Bus.notify(
                    Notification(Configuration.KEY_SWIFTLINT, "Error", "SwiftLint error: $errorLine", NotificationType.INFORMATION)
                )
            }
        }

        log(" --> # Output: \n\t" + outputLines.joinToString(separator = "\n\t"))
        log(" --> # Error: \n\t" + errorLines.joinToString(separator = "\n\t"))

        return outputLines
    }
}